compiler/next/tools/chpldoc/chpldoc.cpp

```diff
@@ -0,0 +1,128 @@
+#include "chpl/queries/Context.h"
+#include "chpl/uast/ASTTypes.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Comment.h"
+
+using namespace chpl;
+using namespace uast;
+using namespace parsing;
+
+// Remove the leading+trailing // or /* (*/)
+std::string stripComment(const std::string &s) {
+  if (s.substr(0, 2) == "//") return s.substr(2);
+  if (s.substr(0, 2) == "/*") {
+    size_t l = s.length();
+    assert(s.substr(l - 2 - 1, 2) == "*/");
+    return s.substr(2, l - 4);
+  }
+  assert(false);
+  return "";
+}
+
+const char *testString = "/* comment 1 */\n"
+                         "module M {\n"
+                         "  // comment 2 \n"
+                         "  x;\n"
+                         "  /* comment 3 */\n"
+                         "}\n"
+                         "/* comment 4 */";
+
+int main(int argc, char **argv) {
+  Context context;
+  Context* ctx = &context;
+
+  auto parser = Parser::build(ctx);
+  Parser* p = parser.get();
+
+  auto parseResult = argc == 2 ?
+    parser->parseFile(argv[1]) :
+    parser->parseString("test1a.chpl", testString);
+
+  if (parseResult.numErrors()) {
+    // for (auto &err : parseResult.errors)
+    for (size_t i = 0; i < parseResult.numErrors(); i++) {
+      // std::cerr << parseResult.error(i) << '\n';
+      auto &err = parseResult.error(i);
+      std::cerr << err.path().c_str() << ":" << err.line() << ":" << err.firstColumn() << ":" << err.message() << "\n";
+    }
+    return 1;
+  }
+
+  std::cerr << "got " << parseResult.numTopLevelExpressions() << " top level expressions\n";
+
+  // TODO why can't I do this
+  // const auto &topLevelExpressions = parseResult.topLevelExpressions();
+  // auto it = std::find_if(topLevelExpressions.begin(), topLevelExpressions.end(), [](const auto &x){ return false; });
+
+  int module_i = -1;
+  for (ssize_t i = 0; i < parseResult.numTopLevelExpressions(); i++) {
+    if (parseResult.topLevelExpression(i)->isModule()) {
+      module_i = i;
+      break;
+    }
+  }
+
+  if (module_i == -1) {
+    std::cerr << "Did not find any top level module expressions\n";
+    return 1;
+  }
+
+  // Print module title
+  const Module *mod = parseResult.topLevelExpression(module_i)->toModule();
+  std::cout << mod->name().c_str() << "\n";
+  std::cout << std::string(mod->name().length(), '=') << "\n\n";
+
+  // Print preceding comment to module
+  if (module_i >= 1) {
+    if (const Comment *comment = parseResult.topLevelExpression(module_i - 1)->toComment()) {
+      std::cout << stripComment(comment->str()) << "\n";
+    }
+  }
+
+  for (size_t i = 0; i < mod->numChildren(); i++) {
+    // TODO things like proc, record, class etc get pulled out into eg
+    // .. {record,function,class,method,itermethod}:: name and signature for functions
+    //    followed by their preceding doc comment
+    //    followed by their attributes and procs
+    if (const Comment *comment = mod->child(i)->toComment()) {
```

Comment:
The compiler offers two ways of applying a visitor pattern which could prove useful for implementing the traversal, see `dispatch` and `traverse` in `include/chpl/uast/ASTNode.h`. You can filter on a parent UAST class, for which `NamedDecl` might be a good candidate (parent class for variables, modules, types, etc - any declaration with a name).

Reply:
I will check those out. At a glance one difficulty here is that the comment I'm interested in is a sibling of the `proc` (for example) and I don't think this needs to go more than 1 level deep. So maybe a `traverse` on the `module` would be the right thing? Will keep that in mind going forward